cbdr/cortex
# COR-710: Publish System Rebuild

## Purpose:
This PR is for a rebuild / reimagining of the Cortex Publish System for created content items. The way this new system works is by using flexibly defined states from the decorator, rather than hard coded state machine states, then determining the current state of a content_item using the relevant date time fields. This is replacing our ActiveJob enqueued approach to handling scheduling and publishing.

Additional inclusions in this PR would be the new names for Employer Blog Categories in the Category tree, which was from a minor task by @justin3thompson, *AND* adds back the expiration date, since this new system makes that concern extremely simple.

> As a content creator, I want my Content to Publish reliably and on-time. Our current system may have a weak spot - currently, it uses our Queuing system, which doesn't enforce reliability, can lose data, and will completely break if you deploy a code change that is incompatible with the version of code that initially spawned the Job.

## JIRA:
https://cb-content-enablement.atlassian.net/browse/COR-710

## Steps to Take On Prod
* Create a new Employer Blog content item
* Schedule the Publish Date to be ~10 minutes in the future (there's a 10 minute cache time)
* Schedule the Expiration Date to be ~10 minutes after that
* Hit Save
* Notice how the state is draft
* Wait until the first 10 minutes have passed, then confirm the state is published (clear the cache if it hasn't updated)
* Wait the next 10 minutes then confirm the state is expired

## Changes:
* Changes to setup
  * N/A

* Architectural changes
  * Rebuilds out publishing system from the ground up

* Migrations
  * N/A

* Library changes
  * Removes our dependency on transitions state_machine, we may wind up removing it altogether

* Side effects
  * N/A

## Screenshots
* Before
N/A

* After
N/A

## QA Links:
http://web.cortex-7.development.c66.me/

## How to Verify These Changes
* Specific pages to visit
  * Create new Employer Blog

* Steps to take
  * Create a new Employer Blog content item
  * Schedule the Publish Date to be ~10 minutes in the future (there's a 10 minute cache time)
  * Schedule the Expiration Date to be ~10 minutes after that
  * Hit Save
  * Notice how the state is draft
  * Wait until the first 10 minutes have passed, then confirm the state is published (clear the cache if it hasn't updated)
  * Wait the next 10 minutes then confirm the state is expired

* Responsive considerations
  * N/A

## Relevant PRs/Dependencies:
N/A

## Additional Information
N/A
